### PR TITLE
feat: migrate appveyor to circleci

### LINKED ISSUE #28241: [CI] Migrate Windows builds to Circle CI
First, some background: The Facebook Open Source team has been working on improving the tooling that supports this open source repository. For example, whenever a Facebook engineer is working on a change to React Native using our internal source control tooling, they can export their patch to a GitHub PR against the open source repository. When they do that, any failures in external CI will be reported alongside our own internal test results.

As it happens, AppVeyor is not widely used at Facebook, and its integration with our internal tooling will no longer be supported in the future. Historically, React Native used AppVeyor because Circle CI did not support Windows. Now that [Circle CI has Windows support](https://circleci.com/docs/2.0/hello-world-windows/), we can consolidate on a single CI again.

If you are interested in working on this, please comment below. You can find the existing AppVeyor config at https://github.com/facebook/react-native/blob/master/.appveyor/config.yml and the Circle CI config at https://github.com/facebook/react-native/blob/master/.circleci/config.yml You can find an example of the work required to do this by looking at the React pull request where they moved away from AppVeyor: https://github.com/facebook/react/pull/17984

### PR DESCRIPTION
**Pull request by Naturalclar**
## Summary

This issue closes #28241
Migrated Windows test from AppVeyor to CircleCI

## Changelog

[Internal] [Changed] - Migrated Windows test from AppVeyor to CircleCI

## Test Plan

For CircleCI to Pass

### CODE REVIEW

**Comment by hramos** [comment]
Thanks for taking this on! Let me know how I can help.

**Comment by gengjiawen** [comment]
Please also add android build on windows. previous we disable it because appveyor  memory is not enough.

**Comment by Naturalclar** [comment]
@hramos thank you! I'm preparing a Windows VM on my pc so I won't have the spam the CI, since I don't own a windows pc 😓, I'll continue working on this soon!

File: .circleci/config.yml [hunk 1]
```diff
@@ -628,6 +636,70 @@ jobs:
...
+      - NDK_TOOLS_URL: https://dl.google.com/android/repository/android-ndk-r19c-windows-x86_64.zip
+    steps:
+      - restore_cache_checkout:
+          checkout_type: windows 
+      - run:
+          name: Install Android SDK Tools
+          command: choco install android-sdk
+      - run:
+          name: Setup SDK
+          command: sdkmanager --licenses > nul
```
Comment (gengjiawen) [inline .circleci/config.yml]:
```suggestion
          command: sdkmanager --licenses
```

Current CI failed here, maybe just remove `> nul`

File: .circleci/config.yml [hunk 2]
```diff
@@ -628,6 +636,70 @@ jobs:
...
+      - run: sdkmanager "system-images;android-19;google_apis;armeabi-v7a"
+      - run: sdkmanager "platforms;android-%ANDROID_BUILD_VERSION%"
+      - run: sdkmanager "build-tools;%ANDROID_TOOLS_VERSION%"
+      - run: sdkmanager "add-ons;addon-google_apis-google-23"
+      - run: sdkmanager "extras;android;m2repository"
+      - run:
```
Comment (gengjiawen) [inline .circleci/config.yml]: ndk can now be installed via `sdkmanager`, also in this way we can lock it's version. See https://github.com/react-native-community/docker-android/blob/a4ae55731f635cc676afd5dec16565b41241355a/Dockerfile#L84.

File: .circleci/config.yml [hunk 3]
```diff
@@ -628,6 +636,70 @@ jobs:
...
+          key: v5-win-node-{{ arch }}-{{ .Branch }}-{{ checksum "yarn.lock" }}
+          paths:
+            - C:\Users\circleci\AppData\Local\Yarn
+      - run:
+          name: Flow Check Android
+          command: yarn flow-check-android
+      - run:
+          name: Flow Check iOS
+          command: yarn flow-check-ios
+      - run:
```
Comment (gengjiawen) [inline .circleci/config.yml]: Consider add `gradlew.bat RNTester:android:app:assembleRelease` here to test windows build.

File: .circleci/config.yml (hunk 2 above)
Comment (Naturalclar) [reply .circleci/config.yml]: Didn't know that, thank you!

File: .circleci/config.yml [hunk 4]
```diff
@@ -628,6 +636,69 @@ jobs:
...
+          name: Test Android Build
+          command: gradlew.bat RNTester:android:app:assembleRelease
```
Comment (gengjiawen) [inline .circleci/config.yml]: if the default is powershell, it's should starts with `./` IIRC.

Comment (Naturalclar) [reply .circleci/config.yml]: I was just about to ask, thanks!

Comment (gengjiawen) [reply .circleci/config.yml]: Also you need to remove `appveyor` badge in `README.md`

Comment (gengjiawen) [reply .circleci/config.yml]: Looks like CI failed on install sdk. Not sure we need to do it, maybe use appveyor provides if they provide. You can use npx envinfo to see this.

**Comment by Naturalclar** [comment]
Pretty close to getting this ready I think.. I just need to find out where the ndk is installed in order to run the android test.

File: .circleci/config.yml [hunk 5]
```diff
@@ -628,6 +636,69 @@ jobs:
...
 
+  # -------------------------
+  #    JOBS: Windows
+  # -------------------------
+  windows_job:
+    executor: 
+      name: win/default
+    environment:
+      - ANDROID_HOME: "C:\\Android\\android-sdk"
+      - ANDROID_NDK: "C:\\Android\\android-sdk\\android-ndk-r20"
```
Comment (gengjiawen) [inline .circleci/config.yml]: NDK should be in `C:\\Android\\android-sdk\\ndk\\20.0.5594570`

**Comment by gengjiawen** [comment]
Looks like on windows build is break, see https://circleci.com/gh/facebook/react-native/139220?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link. @hramos Can you find someone to help ?

**Comment by Naturalclar** [comment]
@gengjiawen thank you so much for all the reviews!

**Comment by hramos** [comment]
Was this an issue in Appveyor?

> make: *** [C:\Users\circleci\project\ReactAndroid\build\tmp\buildReactNdkLib/local/armeabi-v7a/objs/hermes-executor-debug/C_/Users/circleci/project/ReactAndroid/src/main/java/com/facebook/react/__/hermes/reactexecutor/HermesExecutorFactory.o] Error 1
make: *** Waiting for unfinished jobs....
In file included from C:/Users/circleci/project/ReactAndroid/src/main/java/com/facebook/react/../hermes/reactexecutor/OnLoad.cpp:9:
C:/Users/circleci/project/ReactAndroid/src/main/java/com/facebook/react/../hermes/reactexecutor\HermesExecutorFactory.h:10:10: fatal error: 'hermes/hermes.h' file not found
#include <hermes/hermes.h>
         ^~~~~~~~~~~~~~~~~
1 error generated.
make: *** [C:\Users\circleci\project\ReactAndroid\build\tmp\buildReactNdkLib/local/armeabi-v7a/objs/hermes-executor-debug/C_/Users/circleci/project/ReactAndroid/src/main/java/com/facebook/react/__/hermes/reactexecutor/OnLoad.o] Error 1
make: Leaving directory `C:/Users/circleci/project/ReactAndroid/src/main/jni/react/jni'

**Comment by gengjiawen** [comment]
> Was this an issue in Appveyor?
> 

We have disable related build for OOM issue.

**Comment by gengjiawen** [comment]
> > Was this an issue in Appveyor?
> 
> We have disable related build for OOM issue.

Another reason maybe related to ndk version. @Naturalclar Can you try ndk 19 ?

**Comment by Naturalclar** [comment]
@gengjiawen Thanks for the suggestion! Unfortunately, ndk version 19 didn't work as well.. 😢

All the other tests (which was originally in AppVeyor) seemed to have passed :tada:

I think rather than to have this PR stale, it might be better to leave the Android build test on Windows out in this PR as it was originally commented out in AppVeyor as well.
We could add the Android Build Test on Windows under the umbrella in https://github.com/facebook/react-native/issues/23561 for further fixes. How does that sound?

cc: @hramos

**Comment by gengjiawen** [comment]
LGTM. @hramos What do you think ?

**Comment by gengjiawen** [comment]
@Naturalclar Thanks for the contribution :)

**Pull request closed without merging**